File: C++/29_smallest_subarray_with_sum.cpp
/*
Idea:
Iterating through array, storing sum of temporary subarray using 2 pointers,
& moving them if sum is higher or lower ( using sliding window technique)
*/

#include <bits/stdc++.h>
using namespace std;


class Solution {
public:

    int smallestSubWithSum(int arr[], int n, int x)
    {
        int sum = 0, ans = n + 1, i = 0, j = 0;
        while (j < n && i < n)
        {
            //moving front & back pointer till we have sum > target
            while (sum <= x && j < n)
                sum += arr[j++];
            while (sum > x && i < n)
            {
                // checking if subarray is smaller than current one
                ans = min(ans, j - i);
                sum -= arr[i++];
            }
        }
        return ans <= n ? ans : 0;
    }
};


int main() {
    int t;
    cin >> t;
    while (t--)
    {
        int n, x;
        cin >> n >> x;
        int a[n];
        for (int i = 0;i < n;i++)
            cin >> a[i];
        Solution obj;
        cout << obj.smallestSubWithSum(a, n, x) << endl;
    }
    return 0;
}

File: Python/Array/29_smallest_subarray_with_sum.py
"""
Idea:
Iterating through array, storing sum of temporary subarray using 2 pointers,
& moving them if sum is higher or lower ( using sliding window technique)
"""


def smallestSubWithSum(a, n, x):
    total = 0
    ans = n + 1
    i = j = 0
    while i < n and j < n:

        # moving front & back pointer till we have sum > target
        while total <= x and j < n:
            total += a[j]
            j += 1
        while total > x and i < n:
            # checking if subarray is smaller than current one
            ans = min(ans, j - i)
            total -= a[i]
            i += 1

    return 0 if ans > n else ans


def main():
    T = int(input())

    while T > 0:
        sz = [int(x) for x in input().strip().split()]
        n, m = sz[0], sz[1]
        a = [int(x) for x in input().strip().split()]
        print(smallestSubWithSum(a, n, m))

        T -= 1


if __name__ == "__main__":
    main()
